src/control/server/drpc.go

```diff
@@ -164,33 +218,78 @@ func newDrpcCall(method drpc.Method, bodyMessage proto.Message) (*drpc.Call, err
 // makeDrpcCall opens a drpc connection, sends a message with the
 // protobuf message marshalled in the body, and closes the connection.
 // drpc response is returned after basic checks.
-func makeDrpcCall(client drpc.DomainSocketClient, method drpc.Method, body proto.Message) (drpcResp *drpc.Response, err error) {
-	drpcCall, err := newDrpcCall(method, body)
-	if err != nil {
-		return drpcResp, errors.Wrap(err, "build drpc call")
-	}
-
+func makeDrpcCall(ctx context.Context, log logging.Logger, client drpc.DomainSocketClient, method drpc.Method, body proto.Message) (drpcResp *drpc.Response, err error) {
 	client.Lock()
 	defer client.Unlock()
 
-	// Forward the request to the I/O server via dRPC
-	if err = client.Connect(); err != nil {
-		if te, ok := errors.Cause(err).(interface{ Temporary() bool }); ok {
-			if !te.Temporary() {
-				return nil, FaultDataPlaneNotStarted
+	tryCall := func(msg proto.Message) (*drpc.Response, error) {
+		drpcCall, err := newDrpcCall(method, msg)
+		if err != nil {
+			return nil, errors.Wrap(err, "build drpc call")
+		}
+
+		// Forward the request to the I/O server via dRPC
+		if err = client.Connect(); err != nil {
+			if te, ok := errors.Cause(err).(interface{ Temporary() bool }); ok {
+				if !te.Temporary() {
+					return nil, FaultDataPlaneNotStarted
+				}
 			}
+			return nil, errors.Wrap(err, "connect to client")
+		}
+		defer client.Close()
+
+		if drpcResp, err = client.SendMsg(drpcCall); err != nil {
+			return nil, errors.Wrap(err, "send message")
 		}
-		return drpcResp, errors.Wrap(err, "connect to client")
-	}
-	defer client.Close()
 
-	if drpcResp, err = client.SendMsg(drpcCall); err != nil {
-		return drpcResp, errors.Wrap(err, "send message")
+		if err = checkDrpcResponse(drpcResp); err != nil {
+			return nil, errors.Wrap(err, "validate response")
+		}
+
+		return drpcResp, nil
 	}
 
-	if err = checkDrpcResponse(drpcResp); err != nil {
-		return drpcResp, errors.Wrap(err, "validate response")
+	if rdr, ok := isRetryable(body); ok {
+		for {
+			retryable := false
+			drpcResp, err = tryCall(rdr.GetMessage())
+			if err != nil {
+				return nil, err
+			}
+
+			dsr := new(daosStatusResp)
+			if uErr := proto.Unmarshal(drpcResp.Body, dsr); uErr != nil {
+				return
+			}
+			status := drpc.DaosStatus(dsr.Status)
+
+			for _, retryableStatus := range rdr.RetryableStatuses {
+				if status == retryableStatus {
+					retryable = true
+					break
+				}
+			}
+
+			if !retryable {
+				return
+			}
+
+			retryAfter := rdr.RetryAfter
+			if retryAfter == 0 {
+				retryAfter = defaultRetryAfter
+			}
+
+			log.Infof("method %s: retryable %s; retrying after %s", method, status, retryAfter)
```

Comment:
That's nice that the retryAfter time.Duration stringifier handles the units.  Looks good.